File: app/src/main/java/ca/judacribz/gainzassist/activities/start_workout/WorkoutPagerAdapter.java
package ca.judacribz.gainzassist.activities.start_workout;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import java.util.ArrayList;

import android.util.Log;
import ca.judacribz.gainzassist.activities.start_workout.fragments.*;
import ca.judacribz.gainzassist.models.Exercise;
import ca.judacribz.gainzassist.models.Workout;
import org.parceler.Parcels;

public class WorkoutPagerAdapter extends FragmentPagerAdapter {

    // Constants
    // --------------------------------------------------------------------------------------------
    final public static String EXTRA_WARMUPS = "ca.judacribz.gainzassist.activities.start_workout.EXTRA_WARMUPS";
    final public static String EXTRA_MAIN_EXERCISES = "ca.judacribz.gainzassist.activities.start_workout.EXTRA_MAIN_EXERCISES";
    final private static Fragment[] FMTS = new Fragment[] {
            WarmupsList.newInstance(),
            WorkoutScreen.newInstance(),
            ExercisesList.newInstance()
    };
    // --------------------------------------------------------------------------------------------

    // Global Vars
    // --------------------------------------------------------------------------------------------
    private Bundle bundle = new Bundle();
    // --------------------------------------------------------------------------------------------

    // ######################################################################################### //
    // WorkoutPagerAdapter Constructor                                                           //
    // ######################################################################################### //
    @SafeVarargs
    WorkoutPagerAdapter(FragmentManager fragmentManager, ArrayList<Exercise>... exercises) {
        super(fragmentManager);

//        Log.d("WARMUPS", "reps" + warmups.get(0).getSetsList().size());
        bundle.putParcelable(EXTRA_MAIN_EXERCISES, Parcels.wrap(exercises[0]));
        bundle.putParcelable(EXTRA_WARMUPS, Parcels.wrap(exercises[1]));

        for (Fragment fmt : FMTS) {
            fmt.setArguments(bundle);
        }
    }
    // ######################################################################################### //

    // FragmentPagerAdapter Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    /* Returns total number of pages */
    @Override
    public int getCount() {
        return FMTS.length;
    }

    /* Returns the fragment to display for that page */
    @Override
    public Fragment getItem(int position) {
        return FMTS[position];
    }
    //FragmentPagerAdapter//Override///////////////////////////////////////////////////////////////
}

File: app/src/main/java/ca/gainzassist/adapters/WorkoutPagerAdapter.kt
package ca.gainzassist.adapters

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import ca.gainzassist.activities.add_workout.ExEntry
import ca.gainzassist.models.Exercise

class WorkoutPagerAdapter : FragmentPagerAdapter {

    private var fmts: ArrayList<Fragment> = ArrayList()
    private var numExs = 0
    private var baseId = 0

    constructor(fragmentManager: FragmentManager, numExs: Int) : super(fragmentManager) {
        this.numExs = numExs
        fmts = ArrayList()
        newEntries()
    }

    override fun getCount(): Int {
        return fmts.size
    }

    override fun getItem(position: Int): Fragment {
        return fmts[position]
    }

    fun addTab() {
        numExs++
        newEntry(count)
    }

    fun newEntries() {
        for (i in 0 until numExs) {
            newEntry(i)
        }
    }

    fun newEntry(index: Int) {
        val fmt = ExEntry()
        fmt.setInd(index)
        fmts.add(fmt)
    }

    fun removeFragment(index: Int, exercises: ArrayList<Exercise>) {
        fmts.removeAt(index)
        numExs--
        notifyChangeInPosition()
        var i: Int
        var exercise: Exercise
        for (fmt in fmts) {
            i = fmts.indexOf(fmt)
            val newFmt = ExEntry()
            newFmt.setInd(i)
            exercise = exercises[i]
            if (exercise.name != null) {
                newFmt.updateExFields(exercise)
            }
            fmts[i] = newFmt
        }
    }

    override fun getItemPosition(`object`: Any): Int {
        return POSITION_NONE
    }

    override fun getItemId(position: Int): Long {
        return (baseId + position).toLong()
    }

    private fun notifyChangeInPosition() {
        baseId += count + 1
    }

    fun hideDelete() {
        (fmts[0] as ExEntry).hideDelete()
    }
}
